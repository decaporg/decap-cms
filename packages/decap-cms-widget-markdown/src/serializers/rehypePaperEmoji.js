/**
 * Dropbox Paper outputs emoji characters as images, and stores the actual
 * emoji character in a `data-emoji-ch` attribute on the image. This plugin
 * replaces the images with the emoji characters.
 */
export default function rehypePaperEmoji() {
  function transform(node) {
    if (node.tagName === 'img' && node.properties.dataEmojiCh) {
      return { type: 'text', value: node.properties.dataEmojiCh };
    }
    node.children = node.children ? node.children.map(transform) : node.children;
    return node;
  }

  return transform;
}
