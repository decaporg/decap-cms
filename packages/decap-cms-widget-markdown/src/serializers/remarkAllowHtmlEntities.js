export default function remarkAllowHtmlEntities() {
  this.Parser.prototype.inlineTokenizers.text = text;

  /**
   * This is a port of the `remark-parse` text tokenizer, adapted to exclude
   * HTML entity decoding.
   */
  function text(eat, value, silent) {
    var self = this;
    var methods;
    var tokenizers;
    var index;
    var length;
    var subvalue;
    var position;
    var tokenizer;
    var name;
    var min;

    /* istanbul ignore if - never used (yet) */
    if (silent) {
      return true;
    }

    methods = self.inlineMethods;
    length = methods.length;
    tokenizers = self.inlineTokenizers;
    index = -1;
    min = value.length;

    while (++index < length) {
      name = methods[index];

      if (name === 'text' || !tokenizers[name]) {
        continue;
      }

      tokenizer = tokenizers[name].locator;

      if (!tokenizer) {
        eat.file.fail('Missing locator: `' + name + '`');
      }

      position = tokenizer.call(self, value, 1);

      if (position !== -1 && position < min) {
        min = position;
      }
    }

    subvalue = value.slice(0, min);

    eat(subvalue)({
      type: 'text',
      value: subvalue,
    });
  }
}
