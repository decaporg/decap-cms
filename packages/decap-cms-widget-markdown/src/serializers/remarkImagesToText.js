/**
 * Images must be parsed as shortcodes for asset proxying. This plugin converts
 * MDAST image nodes back to text to allow shortcode pattern matching. Note that
 * this transformation only occurs for images that are the sole child of a top
 * level paragraph - any other image is left alone and treated as an inline
 * image.
 */
export default function remarkImagesToText() {
  return transform;

  function transform(node) {
    const children = node.children.map(child => {
      if (
        child.type === 'paragraph' &&
        child.children.length === 1 &&
        child.children[0].type === 'image'
      ) {
        const { alt, url, title } = child.children[0];
        const value = `![${alt || ''}](${url || ''}${title ? ` "${title}"` : ''})`;
        child.children = [{ type: 'text', value }];
      }
      return child;
    });
    return { ...node, children };
  }
}
