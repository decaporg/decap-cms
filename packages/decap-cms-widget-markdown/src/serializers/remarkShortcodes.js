export function remarkParseShortcodes({ plugins }) {
  const Parser = this.Parser;
  const tokenizers = Parser.prototype.blockTokenizers;
  const methods = Parser.prototype.blockMethods;

  tokenizers.shortcode = createShortcodeTokenizer({ plugins });

  methods.unshift('shortcode');
}

export function getLinesWithOffsets(value) {
  const SEPARATOR = '\n\n';
  const splitted = value.split(SEPARATOR);
  const trimmedLines = splitted
    .reduce(
      (acc, line) => {
        const { start: previousLineStart, originalLength: previousLineOriginalLength } =
          acc[acc.length - 1];

        return [
          ...acc,
          {
            line: line.trimEnd(),
            start: previousLineStart + previousLineOriginalLength + SEPARATOR.length,
            originalLength: line.length,
          },
        ];
      },
      [{ start: -SEPARATOR.length, originalLength: 0 }],
    )
    .slice(1)
    .map(({ line, start }) => ({ line, start }));
  return trimmedLines;
}

function matchFromLines({ trimmedLines, plugin }) {
  for (const { line, start } of trimmedLines) {
    const match = line.match(plugin.pattern);
    if (match) {
      match.index += start;
      return match;
    }
  }
}

function createShortcodeTokenizer({ plugins }) {
  return function tokenizeShortcode(eat, value, silent) {
    // Plugin patterns may rely on `^` and `$` tokens, even if they don't
    // use the multiline flag. To support this, we fall back to searching
    // through each line individually, trimming trailing whitespace and
    // newlines, if we don't initially match on a pattern. We keep track of
    // the starting position of each line so that we can sort correctly
    // across the full multiline matches.
    const trimmedLines = getLinesWithOffsets(value);

    // Attempt to find a regex match for each plugin's pattern, and then
    // select the first by its occurrence in `value`. This ensures we won't
    // skip a plugin that occurs later in the plugin registry, but earlier
    // in the `value`.
    const [{ plugin, match } = {}] = plugins
      .toArray()
      .map(plugin => ({
        match: value.match(plugin.pattern) || matchFromLines({ trimmedLines, plugin }),
        plugin,
      }))
      .filter(({ match }) => !!match)
      .sort((a, b) => a.match.index - b.match.index);

    if (match) {
      if (silent) {
        return true;
      }

      const shortcodeData = plugin.fromBlock(match);

      try {
        return eat(match[0])({
          type: 'shortcode',
          data: { shortcode: plugin.id, shortcodeData },
        });
      } catch (e) {
        console.warn(
          `Sent invalid data to remark. Plugin: ${plugin.id}. Value: ${
            match[0]
          }. Data: ${JSON.stringify(shortcodeData)}`,
        );
        return false;
      }
    }
  };
}

export function createRemarkShortcodeStringifier({ plugins }) {
  return function remarkStringifyShortcodes() {
    const Compiler = this.Compiler;
    const visitors = Compiler.prototype.visitors;

    visitors.shortcode = shortcode;

    function shortcode(node) {
      const { data } = node;
      const plugin = plugins.find(plugin => data.shortcode === plugin.id);
      return plugin.toBlock(data.shortcodeData);
    }
  };
}
