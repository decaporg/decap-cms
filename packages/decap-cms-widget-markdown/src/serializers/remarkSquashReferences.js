import without from 'lodash/without';
import flatten from 'lodash/flatten';
import { u } from 'unist-builder';
import mdastDefinitions from 'mdast-util-definitions';

/**
 * Raw markdown may contain image references or link references. Because there
 * is no way to maintain these references within the Slate AST, we convert image
 * and link references to standard images and links by putting their url's
 * inline. The definitions are then removed from the document.
 *
 * For example, the following markdown:
 *
 * ```
 * ![alpha][bravo]
 *
 * [bravo]: http://example.com/example.jpg
 * ```
 *
 * Yields:
 *
 * ```
 * ![alpha](http://example.com/example.jpg)
 * ```
 *
 */
export default function remarkSquashReferences() {
  return function transformer(tree /*, file */) {
    const getDefinition = mdastDefinitions(tree);


    function transform(getDefinition, node) {
      if (!node) return null;
      /**
       * Bind the `getDefinition` function to `transform` and recursively map all
       * nodes.
       */
      const boundTransform = transform.bind(null, getDefinition);
      const children = node.children ? node.children.map(boundTransform) : node.children;

      /**
       * Combine reference and definition nodes into standard image and link
       * nodes.
       */
      if (['imageReference', 'linkReference'].includes(node.type)) {
        const type = node.type === 'imageReference' ? 'image' : 'link';
        const definition = getDefinition(node.identifier);

        // Extract alt text for imageReference in remark-parse v9
        let alt = undefined;
        if (node.type === 'imageReference') {
          // alt text is a child text node
          if (node.children && node.children.length > 0) {
            alt = node.children.map(child => child.value).join('');
          }
        }

        if (definition) {
          const { title, url } = definition;
          return u(type, { title, url, alt }, children);
        }

        const pre = u('text', node.type === 'imageReference' ? '![' : '[');
        const post = u('text', ']');
        const nodes = children || [u('text', alt)];
        return [pre, ...nodes, post];
      }

      /**
       * Remove definition nodes and filter the resulting null values from the
       * filtered children array.
       */
      if (node.type === 'definition') {
        return null;
      }

      const filteredChildren = without(children, null);

      return { ...node, children: flatten(filteredChildren) };
    }
    const newTree = transform(getDefinition, tree);
    Object.assign(tree, newTree);
  };
}
