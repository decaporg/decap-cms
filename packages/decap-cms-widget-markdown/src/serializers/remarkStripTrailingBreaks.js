import mdastToString from 'mdast-util-to-string';

/**
 * Removes break nodes that are at the end of a block.
 *
 * When a trailing double space or backslash is encountered at the end of a
 * markdown block, Remark will interpret the character(s) literally, as only
 * break entities followed by text qualify as breaks. A manually created MDAST,
 * however, may have such entities, and users of visual editors shouldn't see
 * these artifacts in resulting markdown.
 */
export default function remarkStripTrailingBreaks() {
  function transform(node) {
    if (node.children) {
      node.children = node.children
        .map((child, idx, children) => {
          /**
           * Only touch break nodes. Convert all subsequent nodes to their text
           * value and exclude the break node if no non-whitespace characters
           * are found.
           */
          if (child.type === 'break') {
            const subsequentNodes = children.slice(idx + 1);

            /**
             * Create a small MDAST so that mdastToString can process all
             * siblings as children of one node rather than making multiple
             * calls.
             */
            const fragment = { type: 'root', children: subsequentNodes };
            const subsequentText = mdastToString(fragment);
            return subsequentText.trim() ? child : null;
          }

          /**
           * Always return the child if not a break.
           */
          return child;
        })

        /**
         * Because some break nodes may be excluded, we filter out the resulting
         * null values.
         */
        .filter(child => child)

        /**
         * Recurse through the MDAST by transforming each individual child node.
         */
        .map(transform);
    }
    return node;
  }

  return transform;
}
