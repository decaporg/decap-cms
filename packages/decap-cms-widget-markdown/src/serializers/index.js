import trimEnd from 'lodash/trimEnd';
import unified from 'unified';
import u from 'unist-builder';
import markdownToRemarkPlugin from 'remark-parse';
import remarkToMarkdownPlugin from 'remark-stringify';
import remarkToRehype from 'remark-rehype';
import rehypeToHtml from 'rehype-stringify';
import htmlToRehype from 'rehype-parse';
import rehypeToRemark from 'rehype-remark';

import remarkToRehypeShortcodes from './remarkRehypeShortcodes';
import rehypePaperEmoji from './rehypePaperEmoji';
import remarkAssertParents from './remarkAssertParents';
import remarkPaddedLinks from './remarkPaddedLinks';
import remarkWrapHtml from './remarkWrapHtml';
import remarkToSlate from './remarkSlate';
import remarkSquashReferences from './remarkSquashReferences';
import { remarkParseShortcodes, createRemarkShortcodeStringifier } from './remarkShortcodes';
import remarkEscapeMarkdownEntities from './remarkEscapeMarkdownEntities';
import remarkStripTrailingBreaks from './remarkStripTrailingBreaks';
import remarkAllowHtmlEntities from './remarkAllowHtmlEntities';
import slateToRemark from './slateRemark';
import { getEditorComponents } from '../MarkdownControl';

/**
 * This module contains all serializers for the Markdown widget.
 *
 * The value of a Markdown widget is transformed to various formats during
 * editing, and these formats are referenced throughout serializer source
 * documentation. Below is brief glossary of the formats used.
 *
 * - Markdown {string}
 *   The stringified Markdown value. The value of the field is persisted
 *   (stored) in this format, and the stringified value is also used when the
 *   editor is in "raw" Markdown mode.
 *
 * - MDAST {object}
 *   Also loosely referred to as "Remark". MDAST stands for MarkDown AST
 *   (Abstract Syntax Tree), and is an object representation of a Markdown
 *   document. Underneath, it's a Unist tree with a Markdown-specific schema.
 *   MDAST syntax is a part of the Unified ecosystem, and powers the Remark
 *   processor, so Remark plugins may be used.
 *
 * - HAST {object}
 *   Also loosely referred to as "Rehype". HAST, similar to MDAST, is an object
 *   representation of an HTML document.  The field value takes this format
 *   temporarily before the document is stringified to HTML.
 *
 * - HTML {string}
 *   The field value is stringified to HTML for preview purposes - the HTML value
 *   is never parsed, it is output only.
 *
 * - Slate Raw AST {object}
 *   Slate's Raw AST is a very simple and unopinionated object representation of
 *   a document in a Slate editor. We define our own Markdown-specific schema
 *   for serialization to/from Slate's Raw AST and MDAST.
 */

/**
 * Deserialize a Markdown string to an MDAST.
 */
export function markdownToRemark(markdown, remarkPlugins) {
  const processor = unified()
    .use(markdownToRemarkPlugin, { fences: true, commonmark: true })
    .use(markdownToRemarkRemoveTokenizers, { inlineTokenizers: ['url'] })
    .use(remarkParseShortcodes, { plugins: getEditorComponents() })
    .use(remarkAllowHtmlEntities)
    .use(remarkSquashReferences)
    .use(remarkPlugins);

  /**
   * Parse the Markdown string input to an MDAST.
   */
  const parsed = processor.parse(markdown);

  /**
   * Further transform the MDAST with plugins.
   */
  const result = processor.runSync(parsed);

  return result;
}

/**
 * Remove named tokenizers from the parser, effectively deactivating them.
 */
function markdownToRemarkRemoveTokenizers({ inlineTokenizers }) {
  inlineTokenizers &&
    inlineTokenizers.forEach(tokenizer => {
      delete this.Parser.prototype.inlineTokenizers[tokenizer];
    });
}

/**
 * Serialize an MDAST to a Markdown string.
 */
export function remarkToMarkdown(obj, remarkPlugins) {
  /**
   * Rewrite the remark-stringify text visitor to simply return the text value,
   * without encoding or escaping any characters. This means we're completely
   * trusting the markdown that we receive.
   */
  function remarkAllowAllText() {
    const Compiler = this.Compiler;
    const visitors = Compiler.prototype.visitors;
    visitors.text = node => node.value;
  }

  /**
   * Provide an empty MDAST if no value is provided.
   */
  const mdast = obj || u('root', [u('paragraph', [u('text', '')])]);

  const remarkToMarkdownPluginOpts = {
    commonmark: true,
    fences: true,
    listItemIndent: '1',

    /**
     * Use asterisk for everything, it's the most versatile. Eventually using
     * other characters should be an option.
     */
    bullet: '*',
    emphasis: '*',
    strong: '*',
    rule: '-',
  };

  const processor = unified()
    .use({ settings: remarkToMarkdownPluginOpts })
    .use(remarkEscapeMarkdownEntities)
    .use(remarkStripTrailingBreaks)
    .use(remarkToMarkdownPlugin)
    .use(remarkAllowAllText)
    .use(createRemarkShortcodeStringifier({ plugins: getEditorComponents() }))
    .use(remarkPlugins);

  /**
   * Transform the MDAST with plugins.
   */
  const processedMdast = processor.runSync(mdast);

  /**
   * Serialize the MDAST to markdown.
   */
  const markdown = processor.stringify(processedMdast).replace(/\r?/g, '');

  /**
   * Return markdown with trailing whitespace removed.
   */
  return trimEnd(markdown);
}

/**
 * Convert Markdown to HTML.
 */
export function markdownToHtml(markdown, { getAsset, resolveWidget, remarkPlugins = [] } = {}) {
  const mdast = markdownToRemark(markdown, remarkPlugins);

  const hast = unified()
    .use(remarkToRehypeShortcodes, { plugins: getEditorComponents(), getAsset, resolveWidget })
    .use(remarkToRehype, { allowDangerousHTML: true })
    .runSync(mdast);

  const html = unified()
    .use(rehypeToHtml, {
      allowDangerousHtml: true,
      allowDangerousCharacters: true,
      closeSelfClosing: true,
      entities: { useNamedReferences: true },
    })
    .stringify(hast);

  return html;
}

/**
 * Deserialize an HTML string to Slate's Raw AST. Currently used for HTML
 * pastes.
 */
export function htmlToSlate(html) {
  const hast = unified().use(htmlToRehype, { fragment: true }).parse(html);

  const mdast = unified()
    .use(rehypePaperEmoji)
    .use(rehypeToRemark, { minify: false })
    .runSync(hast);

  const slateRaw = unified()
    .use(remarkAssertParents)
    .use(remarkPaddedLinks)
    .use(remarkWrapHtml)
    .use(remarkToSlate)
    .runSync(mdast);

  return slateRaw;
}

/**
 * Convert Markdown to Slate's Raw AST.
 */
export function markdownToSlate(markdown, { voidCodeBlock, remarkPlugins = [] } = {}) {
  const mdast = markdownToRemark(markdown, remarkPlugins);

  const slateRaw = unified()
    .use(remarkWrapHtml)
    .use(remarkToSlate, { voidCodeBlock })
    .runSync(mdast);

  return slateRaw.children;
}

/**
 * Convert a Slate Raw AST to Markdown.
 *
 * Requires shortcode plugins to parse shortcode nodes back to text.
 *
 * Note that Unified is not utilized for the conversion from Slate's Raw AST to
 * MDAST. The conversion is manual because Unified can only operate on Unist
 * trees.
 */
export function slateToMarkdown(raw, { voidCodeBlock, remarkPlugins = [] } = {}) {
  const mdast = slateToRemark(raw, { voidCodeBlock });
  const markdown = remarkToMarkdown(mdast, remarkPlugins);
  return markdown;
}
