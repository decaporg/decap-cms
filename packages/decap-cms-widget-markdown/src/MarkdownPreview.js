import React from 'react';
import PropTypes from 'prop-types';
import { WidgetPreviewContainer } from 'decap-cms-ui-default';
import DOMPurify from 'dompurify';

import { markdownToHtml } from './serializers';
class MarkdownPreview extends React.Component {
  static propTypes = {
    getAsset: PropTypes.func.isRequired,
    resolveWidget: PropTypes.func.isRequired,
    value: PropTypes.string,
  };

  componentDidMount() {
    // Manually validate PropTypes - React 19 breaking change
    PropTypes.checkPropTypes(MarkdownPreview.propTypes, this.props, 'prop', 'MarkdownPreview');
  }

  render() {
    const { value, getAsset, resolveWidget, field, getRemarkPlugins } = this.props;
    if (value === null) {
      return null;
    }

    const html = markdownToHtml(value, { getAsset, resolveWidget }, getRemarkPlugins?.());
    const toRender = field?.get('sanitize_preview', false) ? DOMPurify.sanitize(html) : html;

    return <WidgetPreviewContainer dangerouslySetInnerHTML={{ __html: toRender }} />;
  }
}

export default MarkdownPreview;
