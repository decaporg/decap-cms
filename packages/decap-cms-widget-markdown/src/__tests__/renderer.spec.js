import React from 'react';
import { render, screen } from '@testing-library/react';
import padStart from 'lodash/padStart';
import { Map } from 'immutable';

import MarkdownPreview from '../MarkdownPreview';
import { markdownToHtml } from '../serializers';

describe('Markdown Preview renderer', () => {
  describe('Markdown rendering', () => {
    describe('General', () => {
      it('should render markdown', async () => {
        const value = `
# H1

Text with **bold** & _em_ elements

## H2

* ul item 1
* ul item 2

### H3

1. ol item 1
1. ol item 2
1. ol item 3

#### H4

[link title](http://google.com)

##### H5

![alt text](https://pbs.twimg.com/profile_images/678903331176214528/TQTdqGwD.jpg)

###### H6

![](https://pbs.twimg.com/profile_images/678903331176214528/TQTdqGwD.jpg)
`;
        const html = await markdownToHtml(value);

        const { container } = render(
          <MarkdownPreview value={html} getAsset={jest.fn()} resolveWidget={jest.fn()} />,
        );
        expect(screen.getByRole('heading', { level: 1 })).toHaveTextContent('H1');
        expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent('H2');
        expect(screen.getByRole('heading', { level: 3 })).toHaveTextContent('H3');
        expect(screen.getByRole('heading', { level: 4 })).toHaveTextContent('H4');
        expect(screen.getByRole('heading', { level: 5 })).toHaveTextContent('H5');
        expect(screen.getByRole('heading', { level: 6 })).toHaveTextContent('H6');
        expect(container).toHaveTextContent('Text with bold & em elements');
        expect(screen.getByRole('link', { name: 'link title' })).toHaveAttribute(
          'href',
          'http://google.com',
        );
        expect(screen.getAllByRole('img').length).toBe(2);
      });
    });

    describe('Headings', () => {
      for (const heading of [...Array(6).keys()]) {
        it(`should render Heading ${heading + 1}`, async () => {
          const value = padStart(' Title', heading + 7, '#');
          const html = await markdownToHtml(value);

          render(<MarkdownPreview value={html} getAsset={jest.fn()} resolveWidget={jest.fn()} />);
          expect(screen.getByRole('heading', { level: heading + 1 })).toHaveTextContent('Title');
        });
      }
    });

    describe('Lists', () => {
      it('should render lists', async () => {
        const value = `
1. ol item 1
1. ol item 2
    * Sublist 1
    * Sublist 2
    * Sublist 3
        1. Sub-Sublist 1
        1. Sub-Sublist 2
        1. Sub-Sublist 3
1. ol item 3
`;
        const html = await markdownToHtml(value);

        const { container } = render(
          <MarkdownPreview value={html} getAsset={jest.fn()} resolveWidget={jest.fn()} />,
        );
        // Check for ordered and unordered lists
        expect(container.querySelectorAll('ol').length).toBeGreaterThan(0);
        expect(container.querySelectorAll('ul').length).toBeGreaterThan(0);
        expect(screen.getByText('ol item 1')).toBeInTheDocument();
        expect(screen.getByText('Sublist 1')).toBeInTheDocument();
        expect(screen.getByText('Sub-Sublist 1')).toBeInTheDocument();
      });
    });

    describe('Links', () => {
      it('should render links', async () => {
        const value = `
I get 10 times more traffic from [Google] than from [Yahoo] or [MSN].

  [Google]: http://google.com/        "Google"
  [Yahoo]: http://search.yahoo.com/  "Yahoo Search"
  [MSN]: http://search.msn.com/    "MSN Search"
`;
        const html = await markdownToHtml(value);

        render(<MarkdownPreview value={html} getAsset={jest.fn()} resolveWidget={jest.fn()} />);
        expect(screen.getByRole('link', { name: 'Google' })).toHaveAttribute(
          'href',
          'http://google.com/',
        );
        expect(screen.getByRole('link', { name: 'Yahoo' })).toHaveAttribute(
          'href',
          'http://search.yahoo.com/',
        );
        expect(screen.getByRole('link', { name: 'MSN' })).toHaveAttribute(
          'href',
          'http://search.msn.com/',
        );
      });
    });

    describe('Code', () => {
      it('should render code', async () => {
        const value = 'Use the `printf()` function.';
        const html = await markdownToHtml(value);

        const { container } = render(
          <MarkdownPreview value={html} getAsset={jest.fn()} resolveWidget={jest.fn()} />,
        );
        expect(container.querySelector('code')).toHaveTextContent('printf()');
      });

      it('should render code 2', async () => {
        const value = '``There is a literal backtick (`) here.``';
        const html = await markdownToHtml(value);

        const { container } = render(
          <MarkdownPreview value={html} getAsset={jest.fn()} resolveWidget={jest.fn()} />,
        );
        expect(container.querySelector('code')).toHaveTextContent(
          'There is a literal backtick (`) here.',
        );
      });
    });

    describe('HTML', () => {
      it('should render HTML as is when using Markdown', async () => {
        const value = `
# Title

<form action="test">
  <label for="input">
    <input type="checkbox" checked="checked" id="input"/> My label
  </label>
  <dl class="test-class another-class" style="width: 100%">
      <dt data-attr="test">Test HTML content</dt>
      <dt>Testing HTML in Markdown</dt>
  </dl>
</form>

<h1 style="display: block; border: 10px solid #f00; width: 100%">Test</h1>
`;
        const html = await markdownToHtml(value);

        const { container } = render(
          <MarkdownPreview value={html} getAsset={jest.fn()} resolveWidget={jest.fn()} />,
        );
        expect(container.querySelector('form')).toBeInTheDocument();
        expect(container.querySelector('dl')).toBeInTheDocument();
        expect(container.querySelector('h1[style]')).toHaveTextContent('Test');
      });
    });
  });

  describe('HTML rendering', () => {
    it('should render HTML', async () => {
      const value = '<p>Paragraph with <em>inline</em> element</p>';
      const html = await markdownToHtml(value);

      const { container } = render(
        <MarkdownPreview value={html} getAsset={jest.fn()} resolveWidget={jest.fn()} />,
      );
      expect(container.querySelector('p')).toHaveTextContent('Paragraph with inline element');
      expect(container.querySelector('em')).toHaveTextContent('inline');
    });
  });

  describe('HTML sanitization', () => {
    it('should sanitize HTML', async () => {
      const value = `<img src="foobar.png" onerror="alert('hello')">`;
      const field = Map({ sanitize_preview: true });

      const { container } = render(
        <MarkdownPreview
          value={value}
          getAsset={jest.fn()}
          resolveWidget={jest.fn()}
          field={field}
        />,
      );
      const img = container.querySelector('img');
      expect(img).toHaveAttribute('src', 'foobar.png');
      expect(img).not.toHaveAttribute('onerror');
    });

    it('should not sanitize HTML', async () => {
      const value = `<img src="foobar.png" onerror="alert('hello')">`;
      const field = Map({ sanitize_preview: false });

      const { container } = render(
        <MarkdownPreview
          value={value}
          getAsset={jest.fn()}
          resolveWidget={jest.fn()}
          field={field}
        />,
      );
      const img = container.querySelector('img');
      expect(img).toHaveAttribute('src', 'foobar.png');
      expect(img).toHaveAttribute('onerror', "alert('hello')");
    });
  });
});
