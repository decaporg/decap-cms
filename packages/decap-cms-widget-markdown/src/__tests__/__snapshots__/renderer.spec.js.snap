// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown Preview renderer HTML rendering should render HTML 1`] = `
.emotion-0 {
  margin: 15px 2px;
}

<div
  className="emotion-0 emotion-1"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<p>Paragraph with <em>inline</em> element</p>",
    }
  }
/>
`;

exports[`Markdown Preview renderer HTML sanitization should not sanitize HTML 1`] = `
.emotion-0 {
  margin: 15px 2px;
}

<div
  className="emotion-0 emotion-1"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<img src=\\"foobar.png\\" onerror=\\"alert('hello')\\">",
    }
  }
/>
`;

exports[`Markdown Preview renderer HTML sanitization should sanitize HTML 1`] = `
.emotion-0 {
  margin: 15px 2px;
}

<div
  className="emotion-0 emotion-1"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<img src=\\"foobar.png\\">",
    }
  }
/>
`;

exports[`Markdown Preview renderer Markdown rendering Code should render code 1`] = `
.emotion-0 {
  margin: 15px 2px;
}

<div
  className="emotion-0 emotion-1"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<p>Use the <code>printf()</code> function.</p>",
    }
  }
/>
`;

exports[`Markdown Preview renderer Markdown rendering Code should render code 2 1`] = `
.emotion-0 {
  margin: 15px 2px;
}

<div
  className="emotion-0 emotion-1"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<p><code>There is a literal backtick (\`) here.</code></p>",
    }
  }
/>
`;

exports[`Markdown Preview renderer Markdown rendering General should render markdown 1`] = `
.emotion-0 {
  margin: 15px 2px;
}

<div
  className="emotion-0 emotion-1"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<h1>H1</h1>
<p>Text with <strong>bold</strong> &amp; <em>em</em> elements</p>
<h2>H2</h2>
<ul>
<li>ul item 1</li>
<li>ul item 2</li>
</ul>
<h3>H3</h3>
<ol>
<li>ol item 1</li>
<li>ol item 2</li>
<li>ol item 3</li>
</ol>
<h4>H4</h4>
<p><a href=\\"http://google.com\\">link title</a></p>
<h5>H5</h5>
<p><img src=\\"https://pbs.twimg.com/profile_images/678903331176214528/TQTdqGwD.jpg\\" alt=\\"alt text\\" /></p>
<h6>H6</h6>
<p><img src=\\"https://pbs.twimg.com/profile_images/678903331176214528/TQTdqGwD.jpg\\" /></p>",
    }
  }
/>
`;

exports[`Markdown Preview renderer Markdown rendering HTML should render HTML as is when using Markdown 1`] = `
.emotion-0 {
  margin: 15px 2px;
}

<div
  className="emotion-0 emotion-1"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<h1>Title</h1>
<form action=\\"test\\">
  <label for=\\"input\\">
    <input type=\\"checkbox\\" checked=\\"checked\\" id=\\"input\\"/> My label
  </label>
  <dl class=\\"test-class another-class\\" style=\\"width: 100%\\">
      <dt data-attr=\\"test\\">Test HTML content</dt>
      <dt>Testing HTML in Markdown</dt>
  </dl>
</form>
<h1 style=\\"display: block; border: 10px solid #f00; width: 100%\\">Test</h1>",
    }
  }
/>
`;

exports[`Markdown Preview renderer Markdown rendering Headings should render Heading 1 1`] = `
.emotion-0 {
  margin: 15px 2px;
}

<div
  className="emotion-0 emotion-1"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<h1>Title</h1>",
    }
  }
/>
`;

exports[`Markdown Preview renderer Markdown rendering Headings should render Heading 2 1`] = `
.emotion-0 {
  margin: 15px 2px;
}

<div
  className="emotion-0 emotion-1"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<h2>Title</h2>",
    }
  }
/>
`;

exports[`Markdown Preview renderer Markdown rendering Headings should render Heading 3 1`] = `
.emotion-0 {
  margin: 15px 2px;
}

<div
  className="emotion-0 emotion-1"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<h3>Title</h3>",
    }
  }
/>
`;

exports[`Markdown Preview renderer Markdown rendering Headings should render Heading 4 1`] = `
.emotion-0 {
  margin: 15px 2px;
}

<div
  className="emotion-0 emotion-1"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<h4>Title</h4>",
    }
  }
/>
`;

exports[`Markdown Preview renderer Markdown rendering Headings should render Heading 5 1`] = `
.emotion-0 {
  margin: 15px 2px;
}

<div
  className="emotion-0 emotion-1"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<h5>Title</h5>",
    }
  }
/>
`;

exports[`Markdown Preview renderer Markdown rendering Headings should render Heading 6 1`] = `
.emotion-0 {
  margin: 15px 2px;
}

<div
  className="emotion-0 emotion-1"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<h6>Title</h6>",
    }
  }
/>
`;

exports[`Markdown Preview renderer Markdown rendering Links should render links 1`] = `
.emotion-0 {
  margin: 15px 2px;
}

<div
  className="emotion-0 emotion-1"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<p>I get 10 times more traffic from <a href=\\"http://google.com/\\" title=\\"Google\\">Google</a> than from <a href=\\"http://search.yahoo.com/\\" title=\\"Yahoo Search\\">Yahoo</a> or <a href=\\"http://search.msn.com/\\" title=\\"MSN Search\\">MSN</a>.</p>",
    }
  }
/>
`;
